xdsl/frontend/python_code_check.py

```diff
@@ -141,3 +145,125 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
+
+            # Similarly, this case (`a: i32 = ...`) can also be redefined as a constant.
+            if isinstance(stmt, ast.AnnAssign) and isinstance(
+                    stmt.target,
+                    ast.Name) and not is_constant(stmt.annotation):
+                defined_variables.add(stmt.target.id)
+                continue
+
+            # This is a constant.
+            if isinstance(stmt, ast.AnnAssign) and is_constant(
+                    stmt.annotation):
+                if not isinstance(stmt.target, ast.Name):
+                    raise CodeGenerationException(
+                        stmt.lineno, stmt.col_offset,
+                        f"All constant expressions have to be assigned to 'ast.Name' nodes."
+                    )
+
+                name = stmt.target.id
+                try:
+                    value = eval(ast.unparse(stmt.value))
+                except Exception:
+                    # TODO: This error message can be improved by matching exact
+                    # exceptions returned by `eval` call.
+                    raise CodeGenerationException(
+                        stmt.lineno, stmt.col_offset,
+                        f"Non-constant expression cannot be assigned to constant variable '{name}' or cannot be evaluated."
+                    )
+
+                # For now, support primitive types only and add a guard to abort
+                # in other cases.
+                if not isinstance(value, int) and not isinstance(value, float):
+                    raise CodeGenerationException(
+                        stmt.lineno, stmt.col_offset,
+                        f"Constant '{name}' has evaluated type '{type(value)}' which is not supported."
+                    )
+
+                # TODO: We should typecheck the value against the type. This
+                # can get tricky since ints can overflow, etc. For example,
+                # `a: Const[i16] = 100000000` should give an error.
+                new_node = ast.Constant(value)
+                inliner = ConstantInliner(name, new_node)
+                for candidate in stmts[(i + 1):]:
+                    inliner.visit(candidate)
+
+                # Ideally, we can prune this AST node now, but it is easier just
+                # to avoid it during code generation phase.
+                continue
+
+            # In case of a function/block definition, we must ensure we process
+            # the nested list of statements as well. Note that if we reached
+            # this then all constants above `i` must have been already inlined.
+            # Hence, it is sufficient to check the function body only.
+            if isinstance(stmt, ast.FunctionDef):
+                new_defined_variables = set(
+                    [arg.arg for arg in stmt.args.args])
+                CheckAndInlineConstants.run_with_variables(
+                    stmt.body, new_defined_variables)
+
+
+@dataclass
+class ConstantInliner(ast.NodeTransformer):
```

Comment:
Could you add documentation on both this class and `CheckAndInlineConstants`?
Just to be clear about what is their roles without having to read their code.